90.1 metrics - EER/IEER

RichSee opened the pull request:
Pull request overview
---------------------
 - Add additional 90.1 (2019) equipment metrics (IEER + Additions for SEER2 & HSPF2)
[ASHRAE 90.1 Metrics for E+ 23-1:Design Doc](https://github.com/RichSee/EnergyPlus/blob/7038-NFP-90-1-Metrics-23-1/design/FY2023/NFP-90-1-Metrics-23-1.md) 
 - see https://github.com/NREL/EnergyPlus/issues/7038

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Brijendra21 committed:
Added blank desing Doc for NFP-90-1-Metrics-23-1

Brijendra21 committed:
Update NFP-90-1-Metrics-23-1.md

Added initial New Feature Proposal

RichSee committed:
Update NFP-90-1-Metrics-23-1.md

Brijendra21 committed:
Update NFP-90-1-Metrics-23-1.md

Minor update to the New Feature Proposal markdown

mjwitte reviewed (commented):
This is similar to the work done in https://github.com/NREL/EnergyPlus/pull/9618. It seems that IEER would also apply to a single-speed `Coil:Cooling:DX:CurveFit:Speed`?

rraustad commented:
> This is similar to the work done in #9618. It seems that IEER would also apply to a single-speed `Coil:Cooling:DX:CurveFit:Speed`?

Yes, for units over 65,000 btu/hr

Brijendra21 committed:
ASHRAE 90.1 metrics updates in E+ 23-1  (#2)

* Update NFP-90-1-Metrics-23-1.md

Added Calculation formulas & Images for IEER - from AHRI 340/360 Standard

* Update NFP-90-1-Metrics-23-1.md

Update to address comments by Mike Witte and Richard Roustad -- about IEER calculation for Coil:Cooling:DX:CurveFit -- for both single stage and multi-stage operating modes.

* Update NFP-90-1-Metrics-23-1.md

Added details for SEER2 & HSPF2 calculations for Coil:Cooling:DX;VariableSpeed, Coil:Heating:DX;VariableSpeed

* Update NFP-90-1-Metrics-23-1.md

updated date

* Update NFP-90-1-Metrics-23-1.md

Editing of formulae and descriptions for SEER2 (Cooling) and HSPF2 (Heating) implementation for the Variable Speed DX coils.

* Update NFP-90-1-Metrics-23-1.md

Added various test files selected to test our implementation.  Subsets may also be used in unit testing.

* Update NFP-90-1-Metrics-23-1.md

Adjust formulas for SEER2 & HSPF2

* Update NFP-90-1-Metrics-23-1.md

SEER2 & HSPF2 calculation changes

* Update NFP-90-1-Metrics-23-1.md

More edits and and updates before posting for review.

* Update NFP-90-1-Metrics-23-1.md

word-smithing

Co-authored-by: Richard A See <RichSee@DigitalAlchemyPro.com>

Brijendra21 committed:
Merge branch 'NREL:develop' into 7038-NFP-90-1-Metrics-23-1

Brijendra21 committed:
Update Test Files list to include WaterCooled condenser examples and to document which test files test the various test conditions. (#3)

* Update NFP-90-1-Metrics-23-1.md

Update Test Files list to include WaterCooled condenser examples and to document which test files test the various test conditions.

* Update NFP-90-1-Metrics-23-1.md

Updated revision date and also added WaterToAirHeatPump equipment type to the list of objects for which IEER will be calculated.

Co-authored-by: Richard A See <RichSee@DigitalAlchemyPro.com>

RichSee committed:
Update NFP-90-1-Metrics-23-1.md

Add comments and questions in the 14-December-22 Technicalities call.

RichSee committed:
Update NFP-90-1-Metrics-23-1.md

Update the dates for original and updates

Brijendra21 committed:
synced with develop

Brijendra21 committed:
Merged latest from develop

Brijendra21 committed:
Merged latest from develop

Brijendra21 committed:
merged latest from develop

Brijendra21 committed:
Update from develop

Brijendra21 committed:
IEER 2022 Calculations

Brijendra21 committed:
clang related

Brijendra21 committed:
more towards clang failures

Brijendra21 committed:
clang fixes

Myoldmopar commented:
@Brijendra21 don't sweat the Clang Format issues.  We have a script in the repo that can fix them automatically.  If you are able to install Clang Format 10, you can just run `./scripts/dev/apply_formatting.sh` from the root of the repo and it will fix them all.  Note that later versions of Clang Format get a little confused with some of the code that looks like template code.  

At the same time, that's like the last thing you need to worry about.  Let Clang Format fail on GitHub actions, and I can make a small commit right before merging it once the work is fully complete.

amirroth commented on src/EnergyPlus/StandardRatings.hh:
```diff
@@ -168,6 +171,28 @@ namespace StandardRatings {
+                                  std::string const &DXCoilName,
+                                  std::string const &DXCoilType,
+                                  int const DXCoilType_Num,
+                                  Array1A<Real64> const RatedTotalCapacity,
```
You are passing all of these arrays by value. If you are not modifying the contents of the array pass it by `const &`. If you are modifying them, pass it in by plain `&`.

Brijendra21 committed:
EER Calculation at high speed for various coild Types

Brijendra21 committed:
Addressed Warnings | Unused Variables

Brijendra21 committed:
Output Table updates

RichSee committed:
Updated EPlus-Out-EIO.tex file to :
1. update the ratings reporting developed in 2010 standards
2. update the ratings reporting developed in 2017 standards
3. add the ratings reporting developed for 2023 standards

RichSee committed:
Change name of 2010 implementation table back to the original name == "DX Cooling Coil Standard Rating Information"

Brijendra21 committed:
Merge latest from develop

Brijendra21 committed:
Output Table Updates

RichSee committed:
Update NFP-90-1-Metrics-23-1.md

Final update to the Design Document -- to correct scope that was achieved and to add notes for things that should be done in the 23-2 cycle.

Brijendra21 committed:
Merged Latest from develop

Myoldmopar commented:
OK, the time has pretty much come for the input file update here.  I'll take a look but may ping you @RichSee @Brijendra21 if needed.

Myoldmopar committed:
Merge remote-tracking branch 'upstream/develop' into 7038-NFP-90-1-Metrics-23-1

Myoldmopar committed:
Clean up build warnings, including odd ones in the unit test

Myoldmopar committed:
Revert kiva permission changes

Myoldmopar committed:
Revert "Revert kiva permission changes"

This reverts commit 39d334806f22b6c5f70d4055fe8974bc4300a4f6.

Myoldmopar committed:
Revert "Revert "Revert kiva permission changes""

This reverts commit aa66983fa4fbe4470bfd3e7da6cf2f79d088d506.

Myoldmopar committed:
Cleanup format and custom check issues

Myoldmopar committed:
Two tweaks to get unit tests closer

Brijendra21 committed:
Fixed 5 of the failing Tests related to 2 Speed calculations.

Brijendra21 committed:
Cleanup and substantial refactoring to simplify the code structure.

Myoldmopar committed:
Update all example files for new fields, modify epfilter to write out new fields

Myoldmopar commented:
@RichSee I updated all the test files and updated the HVACTemplate code to support the new fields.  This is close now, but still needs the actual transition Fortran implementation.  Unfortunately I need to first make a pull request that updates the E+ code to the next version.  I'll do that next, get that merged into develop, pull it into this branch, and then add the transition code.  I think this should be all set by tomorrow.

@mjwitte if you care to look at what I did in epfilter, or anything else in this PR, that would be great.  I think it's pretty close once I get the transition rules done later/tomorrow.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 7038-NFP-90-1-Metrics-23-1

Myoldmopar committed:
Fix missing performance test idf object updates

Myoldmopar committed:
Remove unused variables, apply format

Myoldmopar committed:
Merge branch '7038-NFP-90-1-Metrics-23-1' of github.com:RichSee/EnergyPlus into 7038-NFP-90-1-Metrics-23-1

Myoldmopar commented:
OK, got develop pulled in so this is ready for transition code now.  I also updated the performance test IDFs, which I missed when I was updating the test files yesterday.  I also cleaned up some unused variable warnings.  At this point, it looks like there may still be one build warning on Linux where GCC thinks it has found an array out of bounds issue at compile time.  I don't see it from inspection, but we'll need to resolve it if it persists.  There are lots of expected audit diffs, table diffs, and EIO additions for the new values, but otherwise CI should be clean.  This is getting close.  I should be able to do transition rules this afternoon.

Brijendra21 committed:
Replaced square braces to curly braces for references to foot notes.

Brijendra21 committed:
Replaced curly braces to '#' prefix for references to foot notes.

Myoldmopar committed:
Merge remote-tracking branch 'upstream/develop' into 7038-NFP-90-1-Metrics-23-1

Myoldmopar committed:
Merge branch '7038-NFP-90-1-Metrics-23-1' of github.com:RichSee/EnergyPlus into 7038-NFP-90-1-Metrics-23-1

Myoldmopar committed:
Add transition code for all covered coil types

Myoldmopar commented:
@RichSee OK, transition code is now in place.  I took several of our example files that have various configurations of these coil types, transitioned them from my 23.1 installed examples, and ran them with the new binary in this branch.  They all ran perfectly and the error files were generally as clean as they were before as far as I can tell.  At this point, the only thing I see left is [this build warning](https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-067c082ba5c69540481d093c5d5993a13a384878-PR9756-x86_64-Linux-Ubuntu-22.04-gcc-11.3.html#build) which seems to indicate that the compiler is detecting an array out of bounds at compile time.  I believe it is legitimately finding it, but I can't immediately see where the problem lies.  I'll have to boot up my 22.04 VM and investigate.

Myoldmopar commented:
Indeed CI is clean except for the compiler-identified array bounds error.  I am installing the latest GCC on my Ubuntu 22 instance so that my machine matches the CI machine and will hopefully emit the error.  The last thing is to get some more review on the actual code changes.  @mjwitte if you have interest/time to look at or test out my epfilter and transition changes, it would be appreciated.

Myoldmopar commented:
Literally cannot reproduce the compiler warning.  I tried in a 22.04 machine, nope.  Tried in a 22.04 docker, nope.  Tried on the same VM we use that builds this branch and produces the warnings, nope.  I'm baffled.  I am going to have to try to find it by code inspection.  There's somewhere in the code where an array is being accessed out of bounds, and the new GCC compiler is able to find it...but it won't tell me when I ask for it directly.

Brijendra21 committed:
Fixes the Compiler warning "array subscript -1 is outside array bounds of ‘const int [1]’ [-Warray-bounds]"

Myoldmopar commented on src/EnergyPlus/StandardRatings.cc:
```diff
@@ -1549,21 +1547,18 @@ namespace StandardRatings {
                                                                                         CondenserType,
                                                                                         TSCCapFTemp,
                                                                                         TSRatedTotCap,
-                                                                                        HSCCapFlow, // Only for HIGH SPEED
```
:+1:

mjwitte reviewed (commented):
@Myoldmopar Reviewed epfilter and transition code, tested some files, tested transition on a sample of each affected coil type. Looks good.

Myoldmopar commented on src/EnergyPlus/Coils/CoilCoolingDX.cc:
```diff
@@ -1026,7 +1026,16 @@ void CoilCoolingDX::reportAllStandardRatings(EnergyPlus::EnergyPlusData &state)
                 state, state.dataOutRptPredefined->pdchDXCoolCoilIEERIP, coil.name, coil.performance.standardRatingIEER * ConvFromSIToIP, 2);
             OutputReportPredefined::addFootNoteSubTable(
-                state, state.dataOutRptPredefined->pdstDXCoolCoil, "ANSI/AHRI ratings account for supply air fan heat and electric power.");
+                state,
```
OK some new footnote content, fine.

Myoldmopar commented on src/EnergyPlus/Coils/CoilCoolingDXCurveFitPerformance.cc:
```diff
@@ -583,6 +583,19 @@ void CoilCoolingDXCurveFitPerformance::calcStandardRatings210240(EnergyPlus::Ene
+        Real64 IEER_2022(0.0);
+        Real64 EER_2022(0.0);
+        Real64 NetCoolingCapRated2022(0.0);
+        std::tie(IEER_2022, NetCoolingCapRated2022, EER_2022) =
```
There is a lot of `std::tie` being used.  I would have welcomed structured bindings, but I don't think we use those much in the code either.  This isn't something that is executed every time step, so I don't think it's worthwhile to make changes since it is working.

Myoldmopar commented on src/EnergyPlus/Coils/CoilCoolingDXCurveFitPerformance.cc:
```diff
@@ -583,6 +583,19 @@ void CoilCoolingDXCurveFitPerformance::calcStandardRatings210240(EnergyPlus::Ene
+            StandardRatings::IEERCalulcationCurveFit(state, "Coil:Cooling:DX:CurveFit", this->normalMode);
+
+        this->standardRatingIEER2 = IEER_2022;
+        this->standardRatingEER2 = EER_2022;
```
But I see you then just assign the temporary tied variables back to member variables on the class.  Why create the temporary variables at all?

Brijendra21 replied:
Removed the temporary variables.

Myoldmopar commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -2411,13 +2411,17 @@ void GetDXCoils(EnergyPlusData &state)
-        if (!lNumericBlanks(5)) {
-            thisDXCoil.InternalStaticPressureDrop = Numbers(5);
-            thisDXCoil.RateWithInternalStaticAndFanObject = true;
+        state.dataDXCoils->DXCoil(DXCoilNum).RatedTotCap(1) = Numbers(1);
```
?? Why did you take out the reference variable?  We are working hard to clean up these array usages and replace them with more readable references, and this undoes a bunch of them for some reason.

Brijendra21 replied:
This reference variable(**thisCoil**) was made after we created our branch and I think the auto merge has picked the fully qualified names at places where we've some modifications (from our branch). I've replaced them now. Thank-you for pointing it out.

Myoldmopar commented on src/EnergyPlus/StandardRatings.hh:
```diff
@@ -357,8 +461,8 @@ namespace StandardRatings {
-        int const nsp                                                   // Number of compressor speeds
-    );
+        int const nsp,                                                  // Number of compressor speeds
+        Array1D<DataHeatBalance::RefrigCondenserType> CondenserType);
```
Is this array being copied each time this function is called?  I know it's not going to be called often, but still seems odd.

Brijendra21 replied:
I've updated it now.

Myoldmopar commented on src/EnergyPlus/VariableSpeedCoils.cc:
```diff
@@ -5829,6 +5834,49 @@ namespace VariableSpeedCoils {
+            default:
+                break;
+            }
+            StandardRatings::CalcDXCoilStandardRating(state,
```
The magic numbers in here are ... not great, but I don't think you are to blame for them.  That aside, is there a chance the nested switch block above could go through the default section?  If so, what would DefrostControl be?  Assuming timed is the default, what about 

```c++
StandardRatings::HPdefrostControl DefrostControl = StandardRatings::HPdefrostControl::Timed;
if (varSpeedCoil.DefrostControl == 2) {
  DefrostControl = StandardRatings::HPdefrostControl::OnDemand;
}
```

Brijendra21 replied:
I've updated the switch block to have DefrostControl as Timed in case of default :
![image](https://github.com/NREL/EnergyPlus/assets/78803858/7f46d5cd-42a6-4a15-9137-fc849831bdac)

Myoldmopar reviewed (commented):
Most of the code changes here are expected and fine.  I had a couple comments that are not necessary to address, but could use a reply.  The one thing that I am confused about is why you "undid" all the reference object uses and reverted them back to long repeated array lookups.  I kinda think that should get cleaned back up.

Brijendra21 committed:
Addressed Comments

Myoldmopar commented:
Even on detailed reporting interval, I'm not able to reproduce the ESO diff showing up here.  Was there another branch that recently caused diffs in the VRF input file "VariableRefrigerantFlow_5Zone_wAirloop"?  If so maybe this branch was just behind develop when it ran.  Otherwise I think this is in good shape, but the diffs are big and can't just be accepted.

Brijendra21 committed:
Merged latest from develop

Myoldmopar commented:
And the diffs are gone, great!  OK, I think this is ready to go.  Anyone got anything else here?

rraustad commented:
With 14,000 lines of code this should be merged.

Myoldmopar commented:
Yes, although in fairness, a massive portion of those are just the IDF field insertions.  Anyway, good point, merging.  Thanks @RichSee and @Brijendra21

RichSee commented:
Thank you Richard and Edwin.
We updated this branch to the latest from Develop this morning, in the hope that this would resolve the diffs.  
Happy to see that it did.  
Cheers!

nealkruis commented:
I just merged these changes into my branch addressing #9966. @RichSee, why are there changes to Water-to-Air Heat Pumps? I am fairly certain they are not within scope of AHRI 210/240 or 340/360.
